### PR TITLE
Parity test

### PR DESCRIPTION
**Pull request by kaushikb11**
## What does this PR do?

Add a longer running Parity test & make it a nightly event.

## Before submitting
- [] Was this discussed/approved via a GitHub issue? (not for typos and docs)
- [x] Did you read the [contributor guideline](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/.github/CONTRIBUTING.md), **Pull Request** section?
- [x] Did you make sure your PR does only one thing, instead of bundling different changes together?
- [] Did you make sure to update the documentation with your changes? (if necessary)
- [x] Did you write any new necessary tests? (not for typos and docs)
- [x] Did you verify new and existing tests pass locally with your changes?
- [] Did you update the [CHANGELOG](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/CHANGELOG.md)? (not for typos, docs, test updates, or internal minor changes/refactorings)

## PR review
Anyone in the community is free to review the PR once the tests have passed.
Before you start reviewing make sure you have read [Review guidelines](https://github.com/PyTorchLightning/pytorch-lightning/wiki/Review-guidelines). In short, see the following bullet-list:

 - [x] Is this pull request ready for review? (if not, please submit in draft mode)
 - [x] Check that all items from **Before submitting** are resolved
 - [x] Make sure the title is self-explanatory and the description concisely explains the PR
 - [x] Add labels and milestones (and optionally projects) to the PR so it can be classified

## Did you have fun?
Make sure you had fun coding 🙃

### CODE REVIEW

**Comment by tchaton** [comment]
Hey @kaushikb11,

Any updates on this one ?

Best,
T.C

**Comment by kaushikb11** [comment]
@tchaton The Job has crossed the 45 mins limit, let me trim some stuff down.

**Comment by Borda** [comment]
> Job has crossed the 45 mins limit, let me trim some stuff down.

I think the time is fine, we want to have it as nightly run not for each PR

**Review by Borda — CHANGES_REQUESTED** [review]
it shall be Azure pipeline and pls name it as `gpu_benchmark.yml`

**Comment by kaushikb11** [comment]
@Borda Updated!

File: tests/helpers/advanced_models.py [hunk 1]
```diff
@@ -228,3 +233,41 @@ def train_dataloader(self):
...
+        loss = F.cross_entropy(y_hat, y)
+        return {'loss': loss}
+
+    def configure_optimizers(self):
+        return torch.optim.Adam(self.parameters(), lr=self.learning_rate)
+
+    def train_dataloader(self):
+        return DataLoader(
+            CIFAR10(
+                root="./",
```
Comment (awaelchli) [inline tests/helpers/advanced_models.py:268]: isn't there a cache folder somewhere that we keep? Maybe I misremember but we had something for caching mnist once 
cc @Borda

**Review by awaelchli — APPROVED** [review]

File: .azure-pipelines/gpu_benchmark.yml [hunk 2]
```diff
@@ -0,0 +1,26 @@
...
+    timeoutInMinutes: 120
+
+    cancelTimeoutInMinutes: 2
+
+    pool: gridai-spot-pool
+
+    container:
+      # base ML image: mcr.microsoft.com/azureml/openmpi3.1.2-cuda10.2-cudnn8-ubuntu18.04
+      image: "pytorchlightning/pytorch_lightning:base-cuda-py3.8-torch1.6"
+      options: "--runtime=nvidia -e NVIDIA_VISIBLE_DEVICES=all"
```
Comment (carmocca) [inline .azure-pipelines/gpu_benchmark.yml:18]: Note that these are not the same options and torch version we have in the GPU tests.

Is that intended?

**Review by carmocca — COMMENTED** [review]
Since this won't be part of the PR pipeline, how will we know this started failing if it does?

File: .azure-pipelines/gpu-tests.yml [hunk 3]
```diff
@@ -104,10 +104,6 @@ jobs:
         testRunTitle: '$(Agent.OS) - $(Build.BuildNumber)[$(Agent.JobName)] - Python $(python.version)'
       condition: succeededOrFailed()
 
-    - bash: |
```
Comment (Borda) [inline .azure-pipelines/gpu-tests.yml:107]: I think we still shall run some small benchmark or you think the small is too small so it does not make sense?

File: benchmarks/test_basic_parity.py [hunk 4]
```diff
@@ -43,20 +43,16 @@ def assert_parity_absolute(pl_values, pt_values, norm_by: float = 1, max_diff: f
...
 # ParityModuleMNIST runs with num_workers=1
 @pytest.mark.parametrize(
-    'cls_model,max_diff_speed,max_diff_memory',
+    'cls_model,max_diff_speed,max_diff_memory,num_epochs,num_runs',
     [
-        (ParityModuleRNN, 0.05, 0.0),
-        (ParityModuleMNIST, 0.25, 0.0),  # todo: lower this thr
+        (ParityModuleRNN, 0.05, 0.0, 4, 3),
+        (ParityModuleMNIST, 0.25, 0.0, 4, 3),  # todo: lower this thr
+        (ParityModuleCIFAR, 4.0, 0.0002, 2, 2),
```
Comment (Borda) [inline benchmarks/test_basic_parity.py:50]:
```suggestion
        (ParityModuleCIFAR, 0.4, 0.0002, 2, 2),
```
Does that sound as we tolerate 4x slower?

Comment (kaushikb11) [reply benchmarks/test_basic_parity.py:50]: It is absolute! It is 4 seconds

File: .azure-pipelines/gpu-tests.yml (hunk 3 above)
Comment (kaushikb11) [reply .azure-pipelines/gpu-tests.yml]: Updated!

File: .azure-pipelines/gpu_benchmark.yml [hunk 5]
```diff
@@ -0,0 +1,27 @@
...
+
+    workspace:
+      clean: all
+
+    steps:
+      - bash: |
+          export PL_RUNNING_BENCHMARKS=1
+          python -m pytest benchmarks -v --maxfail=2 --durations=0
```
Comment (Borda) [inline .azure-pipelines/gpu_benchmark.yml:25]:
```suggestion
          python -m pytest benchmarks -v --durations=0
```

File: tests/helpers/advanced_models.py (hunk 1 above)
Comment (Borda) [reply tests/helpers/advanced_models.py]: yep, fixed :]

**Review by Borda — APPROVED** [review]

File: .azure-pipelines/gpu_benchmark.yml (hunk 2 above)
Comment (Borda) [reply .azure-pipelines/gpu_benchmark.yml:18]: @kaushikb11 do we need the shared memory option?
cc: @awaelchli

Comment (awaelchli) [reply .azure-pipelines/gpu_benchmark.yml:18]: we needed a little bit more shared memory to run all our GPU tests, but for this parity test the default should be enough. If not, you will see a NCCL error and with `NCCL_DEBUG=INFO` it should say that shared memory is OOM.

File: tests/conftest.py [hunk 6]
```diff
@@ -16,11 +16,19 @@
...
 
 import pytest
 import torch.distributed
 import torch.multiprocessing as mp
 
+from tests import _PATH_DATASETS
+
+
+@pytest.fixture(scope="session")
+def datadir():
```
Comment (awaelchli) [inline tests/conftest.py:29]: nice!

File: .azure-pipelines/gpu_benchmark.yml (hunk 2 above)
Comment (kaushikb11) [reply .azure-pipelines/gpu_benchmark.yml:18]: Got it! I will try removing it.

**Review by carmocca — APPROVED** [review]
